chia/wallet/vault/vault_wallet.py

```diff
@@ -0,0 +1,691 @@
+from __future__ import annotations
+
+import json
+import logging
+import time
+from dataclasses import dataclass
+from typing import Any, Dict, List, Optional, Set, Tuple
+
+from chia_rs import G1Element, G2Element
+from ecdsa.keys import SigningKey
+from typing_extensions import Unpack
+
+from chia.protocols.wallet_protocol import CoinState
+from chia.types.blockchain_format.coin import Coin
+from chia.types.blockchain_format.program import Program
+from chia.types.blockchain_format.sized_bytes import bytes32
+from chia.types.coin_spend import CoinSpend, make_spend
+from chia.types.signing_mode import SigningMode
+from chia.types.spend_bundle import SpendBundle
+from chia.util.hash import std_hash
+from chia.util.ints import uint32, uint64, uint128
+from chia.wallet.coin_selection import select_coins
+from chia.wallet.conditions import Condition, CreateCoin, CreatePuzzleAnnouncement, parse_timelock_info
+from chia.wallet.derivation_record import DerivationRecord
+from chia.wallet.lineage_proof import LineageProof
+from chia.wallet.payment import Payment
+from chia.wallet.puzzles.p2_conditions import puzzle_for_conditions, solution_for_conditions
+from chia.wallet.signer_protocol import (
+    PathHint,
+    SignedTransaction,
+    SigningInstructions,
+    SigningResponse,
+    SigningTarget,
+    Spend,
+    SumHint,
+    TransactionInfo,
+)
+from chia.wallet.transaction_record import TransactionRecord
+from chia.wallet.util.compute_hints import compute_spend_hints_and_additions
+from chia.wallet.util.transaction_type import TransactionType
+from chia.wallet.util.tx_config import CoinSelectionConfig, TXConfig
+from chia.wallet.util.wallet_sync_utils import fetch_coin_spend
+from chia.wallet.util.wallet_types import WalletIdentifier
+from chia.wallet.vault.vault_drivers import (
+    construct_p2_delegated_secp,
+    construct_vault_merkle_tree,
+    get_p2_singleton_puzzle,
+    get_p2_singleton_puzzle_hash,
+    get_recovery_finish_puzzle,
+    get_recovery_inner_puzzle,
+    get_recovery_puzzle,
+    get_recovery_solution,
+    get_vault_full_puzzle,
+    get_vault_full_solution,
+    get_vault_hidden_puzzle_with_index,
+    get_vault_inner_puzzle,
+    get_vault_inner_puzzle_hash,
+    get_vault_inner_solution,
+    get_vault_proof,
+    match_vault_puzzle,
+)
+from chia.wallet.vault.vault_info import RecoveryInfo, VaultInfo
+from chia.wallet.vault.vault_root import VaultRoot
+from chia.wallet.wallet import Wallet
+from chia.wallet.wallet_info import WalletInfo
+from chia.wallet.wallet_protocol import GSTOptionalArgs
+
+
+@dataclass
+class Vault(Wallet):
+    _vault_info: Optional[VaultInfo] = None
+
+    @property
+    def vault_info(self) -> VaultInfo:
+        if self._vault_info is None:
+            raise ValueError("VaultInfo is not set")
+        return self._vault_info
+
+    @vault_info.setter
```

Comment:
I don't know if we want this ergonomic, I think the vault info should probably only be messed with by this class.

Reply:
Do you mean to just set `self._vault_info` directly from `save_info`?

Reply:
Yeah exactly.  We don't want to give the impression that things external to the class can set that attribute.